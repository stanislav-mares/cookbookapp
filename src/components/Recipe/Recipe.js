import React from 'react';
import PropTypes from 'prop-types';

//components
import NewRecipe from 'components/NewRecipe/NewRecipe';

//images
import defaultAvatar from 'images/cookBook.png';
import likeIcon from 'images/likeIcon.png';
import dislikeIcon from 'images/dislikeIcon.png';

export default class Recipe extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      id : this.props.id,
      show: true
    };

    this.likes = 0;
    this.dislikes = 0;
  }

  handleRemove = () => {
    this.props.handleRemoveRecipe(this.state.id);
  }

  handleEditRecipe = (data) => {
    this.props.handleEditRecipe(this.state.id, data);
  }

  handleLikesDislikes = (event) => {

    const name = event.target.name;
    const count = (name === 'likes' ? ++this.likes : ++this.dislikes);

    this.props.handleEditRecipe(this.state.id, {[name] : count})
  }

  render() {

    let ingredients = this.props.ingredients.map((ingredient, key) => {
        return(
          <li key={key}>{ingredient}</li>
        );
    });

    let procedure = this.props.procedure.map((procedure_step, key) => {
        return(
          <p key={key}>{procedure_step}</p>
        );
    });

    return(
      <div>
        <img src={this.props.avatarURL ? this.props.avatarURL : defaultAvatar}
          width="100px" height="100px" alt="Recipe default avatar" />
        <div>
          <h3>{this.props.title}</h3>
          <p>{this.props.desc}</p>
        </div>
        <div>
          <img
            src={likeIcon}
            onClick={this.handleLikesDislikes}
            name="likes" width="15px" height="15px" alt="Like button"
          />
          <span name="likes">{this.props.likes}</span>
          <img
            src={dislikeIcon}
            onClick={this.handleLikesDislikes}
            name="dislikes" width="15px" height="15px" alt="Dislike button"
          />
          <span name="dislikes">{this.props.dislikes}</span>
        </div>
        <div>
          <button onClick={this.handleRemove}>X</button>
        </div>
        <ol>
          {ingredients}
        </ol>
        <div>
          {procedure}
        </div>
        <NewRecipe
          title={this.props.title}
          desc={this.props.desc}
          ingredients={this.props.ingredients}
          procedure={this.props.procedure}
          avatarURL={this.props.avatarURL}
          likes={this.props.likes}
          dislikes={this.props.dislikes}
          show={this.state.show}
          handleRecipe={this.handleEditRecipe}
        />
      </div>
    );
  }
}

Recipe.propTypes = {
  id : PropTypes.number,
  title : PropTypes.string,
  desc : PropTypes.string,
  procedure : PropTypes.arrayOf(PropTypes.string),
  ingredients : PropTypes.arrayOf(PropTypes.string),
  likes: PropTypes.number,
  dislikes: PropTypes.number,
  handleRemoveRecipe: PropTypes.func.isRequired,
  handleEditRecipe: PropTypes.func.isRequired
};
