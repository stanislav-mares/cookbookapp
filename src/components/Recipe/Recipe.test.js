import React from 'react';
import enzymeConfig from 'testConfig/enzymeConfig';
import { shallow } from 'enzyme';

import Recipe from 'components/Recipe/Recipe';

describe('<Recipe />', () => {
  it('testing if likes and dislikes increments after press respectively button', () => {
    const recipe = {
      id: 1,
      title : "Spaghetti",
      desc : "Preparation time: 30min, Difficulty: 3/5",
      ingredients : ["one", "two", "three", "four", "five"],
      procedure: ["step_one", "step_two", "step_three", "step_four", "step_five"],
      avatarURL: '',
      likes : 0,
      dislikes : 0
    };

    const component = shallow(
      <Recipe
        { ...recipe }
        handleEditRecipe={() => {}}
        handleRemoveRecipe={() => {}}
      />
    );
    const instance = component.instance()

    const likesBtn = component.find('img[name="likes"]');
    const dislikesBtn = component.find('img[name="dislikes"]');

    likesBtn.simulate('click', {target:{name:'likes'}});
    likesBtn.simulate('click', {target:{name:'likes'}});
    likesBtn.simulate('click', {target:{name:'likes'}});

    dislikesBtn.simulate('click', {target:{name:'dislikes'}});
    dislikesBtn.simulate('click', {target:{name:'dislikes'}});

    expect(instance.likes).toBe(3);
    expect(instance.dislikes).toBe(2);
  });

  it('testing handleLikesDislikes method of Recipe component', () => {
    const recipe = {
      id: 1,
      title : "Spaghetti",
      desc : "Preparation time: 30min, Difficulty: 3/5",
      ingredients : ["one", "two", "three", "four", "five"],
      procedure: ["step_one", "step_two", "step_three", "step_four", "step_five"],
      avatarURL: '',
      likes : 4,
      dislikes : 4
    };

    const component = shallow(
      <Recipe
        { ...recipe }
        handleEditRecipe={() => {}}
        handleRemoveRecipe={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
