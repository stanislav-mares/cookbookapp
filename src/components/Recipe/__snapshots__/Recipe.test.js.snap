// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe /> testing handleLikesDislikes method of Recipe component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Recipe
    avatarURL=""
    desc="Preparation time: 30min, Difficulty: 3/5"
    dislikes={4}
    handleEditRecipe={[Function]}
    handleRemoveRecipe={[Function]}
    id={1}
    ingredients={
      Array [
        "one",
        "two",
        "three",
        "four",
        "five",
      ]
    }
    likes={4}
    procedure={
      Array [
        "step_one",
        "step_two",
        "step_three",
        "step_four",
        "step_five",
      ]
    }
    title="Spaghetti"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Recipe default avatar"
          height="100px"
          src="test-file-stub"
          width="100px"
        />,
        <div>
          <h3>
            Spaghetti
          </h3>
          <p>
            Preparation time: 30min, Difficulty: 3/5
          </p>
        </div>,
        <div>
          <img
            alt="Like button"
            height="15px"
            name="likes"
            onClick={[Function]}
            src="test-file-stub"
            width="15px"
          />
          <span
            name="likes"
          >
            4
          </span>
          <img
            alt="Dislike button"
            height="15px"
            name="dislikes"
            onClick={[Function]}
            src="test-file-stub"
            width="15px"
          />
          <span
            name="dislikes"
          >
            4
          </span>
        </div>,
        <div>
          <button
            onClick={[Function]}
          >
            X
          </button>
        </div>,
        <ol>
          <li>
            one
          </li>
          <li>
            two
          </li>
          <li>
            three
          </li>
          <li>
            four
          </li>
          <li>
            five
          </li>
        </ol>,
        <div>
          <p>
            step_one
          </p>
          <p>
            step_two
          </p>
          <p>
            step_three
          </p>
          <p>
            step_four
          </p>
          <p>
            step_five
          </p>
        </div>,
        <NewRecipe
          avatarURL=""
          desc="Preparation time: 30min, Difficulty: 3/5"
          dislikes={4}
          handleRecipe={[Function]}
          ingredients={
            Array [
              "one",
              "two",
              "three",
              "four",
              "five",
            ]
          }
          likes={4}
          procedure={
            Array [
              "step_one",
              "step_two",
              "step_three",
              "step_four",
              "step_five",
            ]
          }
          show={true}
          title="Spaghetti"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "alt": "Recipe default avatar",
          "height": "100px",
          "src": "test-file-stub",
          "width": "100px",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              Spaghetti
            </h3>,
            <p>
              Preparation time: 30min, Difficulty: 3/5
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Spaghetti",
            },
            "ref": null,
            "rendered": "Spaghetti",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Preparation time: 30min, Difficulty: 3/5",
            },
            "ref": null,
            "rendered": "Preparation time: 30min, Difficulty: 3/5",
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="Like button"
              height="15px"
              name="likes"
              onClick={[Function]}
              src="test-file-stub"
              width="15px"
            />,
            <span
              name="likes"
            >
              4
            </span>,
            <img
              alt="Dislike button"
              height="15px"
              name="dislikes"
              onClick={[Function]}
              src="test-file-stub"
              width="15px"
            />,
            <span
              name="dislikes"
            >
              4
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "alt": "Like button",
              "height": "15px",
              "name": "likes",
              "onClick": [Function],
              "src": "test-file-stub",
              "width": "15px",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": 4,
              "name": "likes",
            },
            "ref": null,
            "rendered": 4,
            "type": "span",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "alt": "Dislike button",
              "height": "15px",
              "name": "dislikes",
              "onClick": [Function],
              "src": "test-file-stub",
              "width": "15px",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": 4,
              "name": "dislikes",
            },
            "ref": null,
            "rendered": 4,
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
          >
            X
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "X",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "X",
          "type": "button",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <li>
              one
            </li>,
            <li>
              two
            </li>,
            <li>
              three
            </li>,
            <li>
              four
            </li>,
            <li>
              five
            </li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "one",
            },
            "ref": null,
            "rendered": "one",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "two",
            },
            "ref": null,
            "rendered": "two",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": "three",
            },
            "ref": null,
            "rendered": "three",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "host",
            "props": Object {
              "children": "four",
            },
            "ref": null,
            "rendered": "four",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "host",
            "props": Object {
              "children": "five",
            },
            "ref": null,
            "rendered": "five",
            "type": "li",
          },
        ],
        "type": "ol",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              step_one
            </p>,
            <p>
              step_two
            </p>,
            <p>
              step_three
            </p>,
            <p>
              step_four
            </p>,
            <p>
              step_five
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "step_one",
            },
            "ref": null,
            "rendered": "step_one",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "step_two",
            },
            "ref": null,
            "rendered": "step_two",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": "step_three",
            },
            "ref": null,
            "rendered": "step_three",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "host",
            "props": Object {
              "children": "step_four",
            },
            "ref": null,
            "rendered": "step_four",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "host",
            "props": Object {
              "children": "step_five",
            },
            "ref": null,
            "rendered": "step_five",
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "avatarURL": "",
          "desc": "Preparation time: 30min, Difficulty: 3/5",
          "dislikes": 4,
          "handleRecipe": [Function],
          "ingredients": Array [
            "one",
            "two",
            "three",
            "four",
            "five",
          ],
          "likes": 4,
          "procedure": Array [
            "step_one",
            "step_two",
            "step_three",
            "step_four",
            "step_five",
          ],
          "show": true,
          "title": "Spaghetti",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Recipe default avatar"
            height="100px"
            src="test-file-stub"
            width="100px"
          />,
          <div>
            <h3>
              Spaghetti
            </h3>
            <p>
              Preparation time: 30min, Difficulty: 3/5
            </p>
          </div>,
          <div>
            <img
              alt="Like button"
              height="15px"
              name="likes"
              onClick={[Function]}
              src="test-file-stub"
              width="15px"
            />
            <span
              name="likes"
            >
              4
            </span>
            <img
              alt="Dislike button"
              height="15px"
              name="dislikes"
              onClick={[Function]}
              src="test-file-stub"
              width="15px"
            />
            <span
              name="dislikes"
            >
              4
            </span>
          </div>,
          <div>
            <button
              onClick={[Function]}
            >
              X
            </button>
          </div>,
          <ol>
            <li>
              one
            </li>
            <li>
              two
            </li>
            <li>
              three
            </li>
            <li>
              four
            </li>
            <li>
              five
            </li>
          </ol>,
          <div>
            <p>
              step_one
            </p>
            <p>
              step_two
            </p>
            <p>
              step_three
            </p>
            <p>
              step_four
            </p>
            <p>
              step_five
            </p>
          </div>,
          <NewRecipe
            avatarURL=""
            desc="Preparation time: 30min, Difficulty: 3/5"
            dislikes={4}
            handleRecipe={[Function]}
            ingredients={
              Array [
                "one",
                "two",
                "three",
                "four",
                "five",
              ]
            }
            likes={4}
            procedure={
              Array [
                "step_one",
                "step_two",
                "step_three",
                "step_four",
                "step_five",
              ]
            }
            show={true}
            title="Spaghetti"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "alt": "Recipe default avatar",
            "height": "100px",
            "src": "test-file-stub",
            "width": "100px",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                Spaghetti
              </h3>,
              <p>
                Preparation time: 30min, Difficulty: 3/5
              </p>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Spaghetti",
              },
              "ref": null,
              "rendered": "Spaghetti",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Preparation time: 30min, Difficulty: 3/5",
              },
              "ref": null,
              "rendered": "Preparation time: 30min, Difficulty: 3/5",
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="Like button"
                height="15px"
                name="likes"
                onClick={[Function]}
                src="test-file-stub"
                width="15px"
              />,
              <span
                name="likes"
              >
                4
              </span>,
              <img
                alt="Dislike button"
                height="15px"
                name="dislikes"
                onClick={[Function]}
                src="test-file-stub"
                width="15px"
              />,
              <span
                name="dislikes"
              >
                4
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "alt": "Like button",
                "height": "15px",
                "name": "likes",
                "onClick": [Function],
                "src": "test-file-stub",
                "width": "15px",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": 4,
                "name": "likes",
              },
              "ref": null,
              "rendered": 4,
              "type": "span",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "alt": "Dislike button",
                "height": "15px",
                "name": "dislikes",
                "onClick": [Function],
                "src": "test-file-stub",
                "width": "15px",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": 4,
                "name": "dislikes",
              },
              "ref": null,
              "rendered": 4,
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
            >
              X
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "X",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "X",
            "type": "button",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li>
                one
              </li>,
              <li>
                two
              </li>,
              <li>
                three
              </li>,
              <li>
                four
              </li>,
              <li>
                five
              </li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "one",
              },
              "ref": null,
              "rendered": "one",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "two",
              },
              "ref": null,
              "rendered": "two",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": "three",
              },
              "ref": null,
              "rendered": "three",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "host",
              "props": Object {
                "children": "four",
              },
              "ref": null,
              "rendered": "four",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "host",
              "props": Object {
                "children": "five",
              },
              "ref": null,
              "rendered": "five",
              "type": "li",
            },
          ],
          "type": "ol",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                step_one
              </p>,
              <p>
                step_two
              </p>,
              <p>
                step_three
              </p>,
              <p>
                step_four
              </p>,
              <p>
                step_five
              </p>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "step_one",
              },
              "ref": null,
              "rendered": "step_one",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "step_two",
              },
              "ref": null,
              "rendered": "step_two",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": "step_three",
              },
              "ref": null,
              "rendered": "step_three",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "host",
              "props": Object {
                "children": "step_four",
              },
              "ref": null,
              "rendered": "step_four",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "host",
              "props": Object {
                "children": "step_five",
              },
              "ref": null,
              "rendered": "step_five",
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "avatarURL": "",
            "desc": "Preparation time: 30min, Difficulty: 3/5",
            "dislikes": 4,
            "handleRecipe": [Function],
            "ingredients": Array [
              "one",
              "two",
              "three",
              "four",
              "five",
            ],
            "likes": 4,
            "procedure": Array [
              "step_one",
              "step_two",
              "step_three",
              "step_four",
              "step_five",
            ],
            "show": true,
            "title": "Spaghetti",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
